### PR TITLE
Added linters: cppcheck and cpplint

### PR DESCRIPTION
**Pull request by ahcorde**
Added `cppcheck` and `cpplint` to reach QL2.

The strategy it's to download the `cppcheck` and `cpplint` scripts from `ament_lint` package and run them as tests.

@brawner @Blast545 what do you think?

Signed-off-by: ahcorde <ahcorde@gmail.com>

### CODE REVIEW

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -46,3 +46,44 @@ foreach(GTEST_SOURCE_file ${tests})
...
+if(NOT PYTHONINTERP_FOUND)
+  message(STATUS "No PythonInterp found. Linters will not be executed")
+  return()
+endif()
+
+find_program(WGET_EXE wget)
+if(WGET_EXE)
+  message(STATUS "Found WGet: ${WGET_EXE}")
+  add_custom_target(wget_cppcheck
+                    COMMAND ${WGET_EXE} -q -O cppcheck.py https://raw.githubusercontent.com/ament/ament_lint/master/ament_cppcheck/ament_cppcheck/main.py
```
Comment (Blast545) [inline test/CMakeLists.txt:61]: Does the test work only by downloading the `main.py` file?

File: test/CMakeLists.txt [hunk 2]
```diff
@@ -46,3 +46,44 @@ foreach(GTEST_SOURCE_file ${tests})
...
+                   "cpplint.py"
+                   "--counting=detailed"
+                   "--extensions=cpp,h"
+                   "--linelength=100"
+                   "--filter=-build/c++11,-runtime/references,-whitespace/braces,-whitespace/indent,-whitespace/parens,-whitespace/semicolon"
+           "${CMAKE_SOURCE_DIR}/include/console_bridge/console.h"
+           "${CMAKE_SOURCE_DIR}/src/console.cpp"
+  )
+
+  add_dependencies(console_bridge wget_cppcheck wget_cpplint)
```
Comment (Blast545) [inline test/CMakeLists.txt:85]: Why is this part needed?

Comment (ahcorde) [reply test/CMakeLists.txt:85]: otherwise the wget is not executed.

File: test/CMakeLists.txt (hunk 1 above)
Comment (ahcorde) [reply test/CMakeLists.txt:61]: yes, it's enough

File: test/CMakeLists.txt [hunk 3]
```diff
@@ -46,3 +46,44 @@ foreach(GTEST_SOURCE_file ${tests})
 
    set_tests_properties(${BINARY_NAME} PROPERTIES TIMEOUT 240)
 endforeach()
+
+find_package(PythonInterp)
```
[the remark this answers is not in the document]
Comment (ahcorde) [reply test/CMakeLists.txt:50]: https://github.com/ros/console_bridge/pull/73/commits/46fb587deffd400b75d6776f72781433263d2564

**Comment by scpeters** [comment]
it looks like travis-ci didn't run on the latest commit

do we need to switch to GitHub actions?

**Comment by ahcorde** [comment]
I added an empty commit to re-run CI. Right now [travis-ci is working](https://travis-ci.org/github/ros/console_bridge/builds/686660879) but failing. ( I think it's because of the python version)

**Comment by scpeters** [comment]
it looks like travis is using trusty for CI; feel free to update that to a more recent version

**Comment by brawner** [comment]
This is a really cool PR. Since this is neither a ROS 1 package nor a ROS 2 package (technically), what's the rationale for using ament's linters and configurations? Are there other more appropriate lint configurations for this package's 'style'?

**Comment by ahcorde** [comment]
Ament's linters wrap some utilities such as cppcheck or [cpplint](https://github.com/cpplint/cpplint). I'm downloading these scripts from our packages to use the same version as in ROS 2.

For example cppcheck is integrated in CMake, adding the following line:

```make
set(CMAKE_CXX_CPPCHECK "cppcheck")
```

but the linter will be checked during the compilation process and we want to include this as a specific tests that's why I added the logic.

We can add uncrustify.

**Comment by brawner** [comment]
I guess the question is: why ROS 2 style configurations instead of ROS 1 style?

**Comment by scpeters** [comment]
> I guess the question is: why ROS 2 style configurations instead of ROS 1 style?

I think it could go either way. ROS 2 style sounds fine to me

**Comment by brawner** [comment]
Totally. I don't have a preference one way or the other, but I was worried others might. I just wanted to capture some of the thought behind the choice. This PR cuts down line lengths to 100 from 120 for example.

**Comment by scpeters** [comment]
> Totally. I don't have a preference one way or the other, but I was worried others might.

we can listen for objections, but in the absence of any, I think the ros2 style can be used

**Comment by ahcorde** [comment]
My thought was the following: this request came from ROS 2 Quality levels. I can't remember any ROS 1 packages taking care about linters. That's why I decided use ament's linters. Open to include/remove others

**Comment by ahcorde** [comment]
@scpeters, :construction_worker: travis-ci is working fine :smiley:

File: include/console_bridge/console.h [hunk 4]
```diff
@@ -34,12 +34,12 @@
...
 
-#ifndef CONSOLE_BRIDGE_CONSOLE_
-#define CONSOLE_BRIDGE_CONSOLE_
+#ifndef INCLUDE_CONSOLE_BRIDGE_CONSOLE_H_
+#define INCLUDE_CONSOLE_BRIDGE_CONSOLE_H_
 
 #include <string>
 
-#include "console_bridge_export.h"
+#include "./console_bridge_export.h"
```
Comment (traversaro) [inline include/console_bridge/console.h:42]: Why is this necessary?

Comment (ahcorde) [reply include/console_bridge/console.h:42]: cpplinter shows the following error:

> Include the directory when naming .h files  [build/include] [4]

another workaround could be this one, but then this header will be consider as a C system header:

```
#include <console_bridge_export.h>
```

File: test/CMakeLists.txt [hunk 5]
```diff
@@ -46,3 +46,44 @@ foreach(GTEST_SOURCE_file ${tests})
...
+  )
+
+  add_test(NAME console_bridge_cpplint
+           COMMAND "${PYTHON_EXECUTABLE}"
+                   "cpplint.py"
+                   "--counting=detailed"
+                   "--extensions=cpp,h"
+                   "--linelength=100"
+                   "--filter=-build/c++11,-runtime/references,-whitespace/braces,-whitespace/indent,-whitespace/parens,-whitespace/semicolon"
+           "${CMAKE_SOURCE_DIR}/include/console_bridge/console.h"
```
Comment (traversaro) [inline test/CMakeLists.txt:81]: Nitpick: if you use `CMAKE_SOURCE_DIR`, the tests will fail if the project is included in a bigger project via `add_subdirectory`, for example using FetchContent. Using `PROJECT_SOURCE_DIR` in this context is probably more robust.

File: include/console_bridge/console.h (hunk 4 above)
Comment (traversaro) [reply include/console_bridge/console.h:42]: Ack, thanks for the reply. For reference it seems that that cpplint check is some check relative to Google include guidelines that I do not think really apply to general CMake projects (see https://github.com/ament/ament_lint/blob/master/ament_cpplint/ament_cpplint/cpplint.py#L4485), but if the workaround is working and I don't think it is a big problem.

**Comment by ahcorde** [comment]
@scpeters I think all the comments have been addressed. Feel free to merge it when you consider

File: test/CMakeLists.txt [hunk 6]
```diff
@@ -46,3 +46,44 @@ foreach(GTEST_SOURCE_file ${tests})
 
    set_tests_properties(${BINARY_NAME} PROPERTIES TIMEOUT 240)
 endforeach()
+
+find_package(PythonInterp 3 REQUIRED)
+
+if(NOT PYTHONINTERP_FOUND)
+  message(STATUS "No PythonInterp found. Linters will not be executed")
```
Comment (scpeters) [inline test/CMakeLists.txt:53]: I think at the `STATUS` level, these messages will not be shown during a `colcon build`, though they would be at a `WARNING` level. I think that is fine, but I wanted to confirm this is the style of verbosity that is preferred

Comment (ahcorde) [reply test/CMakeLists.txt]: Changed https://github.com/ros/console_bridge/pull/73/commits/c8efe0116935a0b6cb9a1f2228e04876c2e3aff5
